OTLP registry emitter

jerbly opened the pull request:
First iteration of the new command: `registry emit`. Emits a semantic convention registry as example spans to your OTLP receiver. This may be useful in testing/simulation scenarios.

`--endpoint` can be used to set the receiver's URL. Defaults to `http://localhost:4317`. The OpenTelemetry standard `OTEL_EXPORTER_OTLP_ENDPOINT` env var will override this setting.

With this command:
```shell
weaver registry emit
```
A single trace with all defined spans in the current Semantic Conventions are sent to the receiver:
<img width="1503" alt="image" src="https://github.com/user-attachments/assets/c8639930-e2f8-4e7b-a03d-bee8359e9951" />

Spans:
- [x] Emit all defined spans for the registry in a single trace.
- [x] Use standard OpenTelemetry libraries for real-world compatibility
- [x] Write tests using a custom Exporter
- [x] Generate examples from the resolved examples in the registry
- [x] Use the correct `span_kind`

A future PR will cover: 
- Generate examples for template types (correctness issues prevent this: https://github.com/open-telemetry/semantic-conventions/issues/1740)
- Adhere to `conditionally_required` clause (programmatically doing this, is not possible)
- Emit all defined metrics for the registry.

lquerel commented on src/registry/emit.rs:
```diff
@@ -0,0 +1,440 @@
+        prepare_main_registry(&args.registry, &args.policy, logger.clone(), &mut diag_msgs)?;
+
+    logger.loading(&format!("Emitting registry `{}`", args.registry.registry));
+    let rt = tokio::runtime::Runtime::new().expect("failed to create tokio runtime");
+    rt.block_on(async {
+        let tracer_provider = if args.stdout {
+            logger.mute();
+            init_stdout_tracer_provider()
+        } else {
+            init_tracer_provider().expect("OTLP Tracer Provider must be created")
+        };
+        let _ = global::set_tracer_provider(tracer_provider.clone());
+
+        emit_trace_for_registry(&registry, &args.registry.registry.to_string());
+
+        global::shutdown_tracer_provider();
+    });
```
Regarding the conversation with @jsuereth this morning, I think the creation of the Tokio runtime should be encapsulated in the `weaver_emit` crate mentioned earlier to avoid any async aspects at this level (even though it is honestly quite limited).

jerbly replied:
I have done this now.

lquerel reviewed (changes requested):
Overall, I really like the idea of having a semconv registry-driven generation of OTLP signals! It opens up a lot of interesting possibilities. An immediate use case is the ability to use semconv as a mechanism to decouple the development of an application from the development of the observability infrastructure associated with that application (e.g., creating dashboards even before the application is developed). Telemetry-schema driven simulations is another long-term option :-)

I have a few requests for minor changes, and I would also like to get @jsuereth 's feedback on this PR.

Thanks again, this is really great!

jerbly commented:
Totally agree with your comments @lquerel, and will go ahead with those changes. Thanks!

jerbly commented on src/registry/emit.rs:
```diff
@@ -0,0 +1,440 @@
+    #[command(flatten)]
+    pub diagnostic: DiagnosticArgs,
+
+    /// Write the telemetry to standard output
+    #[arg(long)]
+    stdout: bool,
```
Added the `endpoint` config option.

jerbly commented on src/registry/emit.rs:
```diff
@@ -0,0 +1,440 @@
+fn init_tracer_provider() -> Result<sdktrace::TracerProvider, TraceError> {
+    let exporter = opentelemetry_otlp::SpanExporter::builder()
+        .with_tonic()
+        .with_endpoint("http://localhost:4317")
```
Done.

jerbly commented on src/registry/emit.rs:
```diff
@@ -0,0 +1,440 @@
+
+// For the given attribute, return a name/value pair.
+// Values are generated based on the attribute type and examples where possible.
+fn get_attribute_name_value(attribute: &Attribute) -> KeyValue {
```
I have implemented this now. The design currently is to have a separate module for each signal type.

jerbly commented:
@lquerel - I've now refactored this into a crate and added the `endpoint` config option. This moves out the tokio runtime and dependencies too. Also, the additional dependencies I had for otel are now confined to the `weaver_emit` crate. Oh, and there's a bunch more testing!

Please take another look. :)

lquerel reviewed (approved):
Thanks for making these changes. It looks to me like it’s in a mergeable state.
@jsuereth , could you take a look at this PR to finalize the merge? 
Thanks!

jsuereth commented on crates/weaver_emit/src/spans.rs:
```diff
@@ -0,0 +1,961 @@
+    tracer.in_span("weaver.emit", |cx| {
+        let span = cx.span();
+        span.set_attribute(KeyValue::new(
+            "weaver.registry_path",
```
This may be a circular problem, but can we register this attribute in SEMCONV?

Either that or use the reserved `otel` namespace

jsuereth reviewed (approved):
This looks like a great first step.

Opened a comment on the choice of "weaver tracking" comments, that I think we should try to resolve somehow.

Specifically, see: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/semantic-conventions.md#semantic-conventions and the reserved `otel` namespace. I think we could ALSO use it here to denote tooling-relevant names.

At a minimum, you should pull that string into a const in the file so it's obvious what magic strings we use here.

jerbly commented:
@jsuereth - for the moment I'll pull these into a const with a TODO to fix. Should I make them `otel.weaver.*` for now?
